### DISCUSSION TITLE
ParamRF: a declarative circuit modelling framework

### DISCUSSION THREAD

**Discussion by jhillairet** [scikit-rf/scikit-rf#1345] [Show and tell]
For info, ParamRF, or pmrf, is a declarative circuit modelling framework, by @gvcallen:

> The framework caters for the frequency-domain fitting and simulation of (microwave) circuit models in an efficient, parametric, object-orientated manner. Concepts from several packages are integrated, such as scikit-rf for general syntax and RF concepts, equinox for model building, and jax for high-speed, hardware-accelerated calculations and automatic differentiation.

https://github.com/paramrf/paramrf

see also paper on Arxiv:
https://arxiv.org/pdf/2510.15881

**Comment by arsenovic** [comment]
looks cool!

**Comment by Vinc0110** [comment]
Cool. But where is the vector fitting? :O

**Comment by gvcallen** [comment]
Hi all, sorry only seeing this now - thanks for the mention! Also, thanks very much for scikit-rf - a few of the core implementations and maths in ParamRF are really just adaptions of the ones in here and I do appreciate your team's hard work.

ParamRF has grown quite a bit since this post and now has pretty reasonable documentation if anyone wants to check it out. I've also added a [section](https://paramrf.github.io/paramrf/skrf_comparison.html) on when someone should prefer to use skrf vs pmrf, as they really are two different tools that just share some overlap.

@Vinc0110 There is definitely a case for that haha, however the library initially started out purely to define and fit known, complex circuit models to measured data, as opposed to numerical ones. But definitely could be added in the future!

**Reply by jhillairet** [reply]
By the way, please have a look to [`Circuit.update_networks`](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.circuit.Circuit.update_networks.html#skrf.circuit.Circuit.update_networks), which is a relatively recent method which allows fast update of any (eventually big) `Circuit`, which is in particular useful for optimization (made by @Asachoo).

**Reply by gvcallen** [reply]
@jhillairet thanks for the heads up! I had a look, and its definitely a very nice method for scikit-rf! However ParamRF operates slightly different due to its parametric nature and interop with JAX and `jax.jit`, so `update_networks` specifically isn't relevant in my case, though its definitely a good reference!

If you are interested, I recently [profiled ParamRF vs scikit-rf](https://gvcallen.github.io/paramrf/skrf_comparison/performance.html) for a simple circuit, and got between 5x and 10x speedup for a single forward pass. Since ParamRF uses autodiff, this would mean around 250x-500x speedup for a gradient-based optimization of a 50 parameter circuit.

Just thought I'd share if anyone in scikit-rf is interested in looking deeper into the differences and why I designed ParamRF in the first place!
